sdk/monitor/Azure.Monitor.Ingestion/src/Models/UploadLogsOptions.cs

```diff
@@ -21,5 +28,62 @@ public class UploadLogsOptions
         /// <remarks> In the Upload method, this parameter is not used as the batches are uploaded in sequence. For parallel uploads, if this value is not set the default concurrency will be 5. </remarks>
         /// </summary>
         public int MaxConcurrency { get; set; }
+
+        /// <summary>
+        /// test
+        /// </summary>
+        public event SyncAsyncEventHandler<UploadFailedEventArgs> UploadFailedEventHandler;
+
+        internal ClientDiagnostics _clientDiagnostics;
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="uploadFailedArgs"></param>
+        /// <param name="clientDiagnostics"></param>
+        /// <returns></returns>
+        internal virtual async Task InvokeEvent(UploadFailedEventArgs uploadFailedArgs, ClientDiagnostics clientDiagnostics)
+        {
+            await UploadFailedEventHandler.RaiseAsync(uploadFailedArgs, nameof(LogsIngestionClient), "Upload", clientDiagnostics).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="async"></param>
+        /// <param name="eventArgs"></param>
+        /// <param name="options"></param>
+        internal virtual async Task OnExceptionAsync(bool async, UploadFailedEventArgs eventArgs, UploadLogsOptions options)
+        {
+            try
+            {
+                if (!async)
+                {
+#pragma warning disable AZC0106 // Non-public asynchronous method needs 'async' parameter.
+                    options.InvokeEvent(eventArgs, _clientDiagnostics).EnsureCompleted();
```

Comment:
I suppressed the warning because InvokeEvent does not take in an async param as the boolean parameter is part of the UploadFailedEventArgs eventArgs that is passed in.

Reply:
Is `EnsureCompleted` the right move instead of `GetAwaiter().GetResult()`?  `EnsureCompleted` is more of a debugging aid for us than something where customers might implement this correctly causing sync-over-async.